### PR TITLE
SQL: current/historical identifier modes [closes #126]

### LINKED ISSUE #126: Historical identifiers mode (3w)
### Task scope
* make global/session variable **versioning_identifier_mode**;
* `SHOW CREATE` works with historical table names;
* `SELECT` works with historical table names (depends on #127).

### Description
3 modes of historical SELECT identifiers switched by global/session variable **versioning_identifier_mode**:
* `CURRENT` (mode 1);
* `HISTORICAL_EARLY` (mode 2, lower bound);
* `HISTORICAL` (mode 2, upper bound).

See [Mode 2](https://github.com/tempesta-tech/mariadb/wiki/DDL-Survival#mode-2-select-by-past-names) of DDL traverse.

There is no difference between `HISTORICAL_EARLY` and `HISTORICAL` for **AS OF** queries. For **FROM .. TO**, **BETWEEN .. AND** `HISTORICAL_EARLY` selects lower timestamp of range, `HISTORICAL` selects higher timestamp of range.

Depends on #199

### PR DESCRIPTION
**Pull request by kevgs**

### CODE REVIEW

File: sql/sql_parse.cc [hunk 1]
```diff
@@ -6362,6 +6362,73 @@ mysql_execute_command(THD *thd)
...
+  bool can_deadlock= thd->mdl_context.has_locks();
+  bool error= open_tables_only_view_structure(thd, &tl, can_deadlock);
+  if (!error)
+  {
+    fields.clear();
+    TABLE_SHARE *s= tl.table->s;
+    for (size_t i= 0; i < s->fields; i++)
+    {
+      // TODO try remove strdup()
+      if (fields.append(thd->strdup(s->field[i]->field_name)))
```
Comment (midenok) [inline sql/sql_parse.cc]: Is it leak?

File: sql/sql_parse.cc [hunk 2]
```diff
@@ -6362,6 +6362,73 @@ mysql_execute_command(THD *thd)
...
+        return true;
+    }
+  }
+  close_thread_tables(thd);
+  thd->mdl_context.rollback_to_savepoint(
+      open_tables_backup.mdl_system_tables_svp);
```
Comment (midenok) [inline sql/sql_parse.cc]: Fix formatting here and in all other places. Notify when done.

File: sql/vtmd.cc [hunk 3]
```diff
@@ -549,10 +556,36 @@ VTMD_table::find_archive_name(THD *thd, String &out)
...
+          vtmd.table->field[FLD_NAME]->val_str(&out);
+        else
+          vtmd.table->field[FLD_ARCHIVE_NAME]->val_str(&out);
+
+        if (mode == VERS_IDENT_MODE_HISTORICAL_EARLY)
+          break;
+
+        if (mode == VERS_IDENT_MODE_HISTORICAL)
+        {
+          // loop till the last record
```
Comment (midenok) [inline sql/vtmd.cc]: Empty condition?

File: sql/sql_parse.cc (hunk 1 above)
Comment (kevgs) [reply sql/sql_parse.cc]: No. Optimization opportunity.

File: sql/sql_parse.cc (hunk 2 above)
Comment (kevgs) [reply sql/sql_parse.cc]: Fixed some places. Am I missed something?

File: sql/sys_vars.cc [hunk 4]
```diff
@@ -416,6 +416,16 @@ static Sys_var_enum Sys_vers_alter_history(
        SESSION_VAR(vers_alter_history), CMD_LINE(OPT_ARG),
        vers_alter_history_keywords, DEFAULT(VERS_ALTER_HISTORY_KEEP));
 
+static const char *vers_ident_mode_keywords[]= {"CURRENT", "HISTORICAL_EARLY",
+                                                "HISTORICAL", NULL};
+static Sys_var_enum Sys_vers_ident_mode(
+       "versioning_identifier_mode", "Versioning identifier selection mode. "
+       "CURRENT: current table column names to archive table;"
```
Comment (midenok) [inline sql/sys_vars.cc]: current table and column names

File: sql/sys_vars.cc [hunk 5]
```diff
@@ -416,6 +416,16 @@ static Sys_var_enum Sys_vers_alter_history(
...
+       "HISTORICAL_EARLY: take earlier of a range of archive tables;"
```
Comment (midenok) [inline sql/sys_vars.cc]: historical table and column names, earliest for range

File: sql/sys_vars.cc [hunk 6]
```diff
@@ -416,6 +416,16 @@ static Sys_var_enum Sys_vers_alter_history(
...
+       "HISTORICAL: take later of a range of archive tables;",
```
Comment (midenok) [inline sql/sys_vars.cc]: historical table and column names, latest for range

File: mysql-test/suite/versioning/t/vtmd.test [hunk 7]
```diff
@@ -197,6 +197,54 @@ set versioning_alter_history= survive;
...
+
+set versioning_alter_history = survive;
+
+create or replace table t (a int) with system versioning;
+insert into t values (1);
+set @ts1 = now(6);
+alter table t add column b int;
+update t set a=2;
+set @ts2 = now(6);
+alter table t drop column b;
```
Comment (midenok) [inline mysql-test/suite/versioning/t/vtmd.test]: Test case for rename table?

File: mysql-test/suite/versioning/t/vtmd.test [hunk 8]
```diff
@@ -197,6 +197,54 @@ set versioning_alter_history= survive;
...
+update t set a = 3;
+
+delimiter ~~;
+create or replace procedure select_a_from_range()
+begin
+  select a from t for system_time from '0-0-0' to now(6);
```
Comment (midenok) [inline mysql-test/suite/versioning/t/vtmd.test]: If names are important don't just name `a` and `b`. Pick some distinguishable and easy names like `fire` and `ice`.

File: mysql-test/suite/versioning/t/vtmd.test [hunk 9]
```diff
@@ -197,6 +197,54 @@ set versioning_alter_history= survive;
...
+call select_a_from_range();
+set versioning_identifier_mode = historical;
+call select_a_from_range();
+
+drop database test;
+create database test;
+use test;
+
+set versioning_alter_history = survive;
+set versioning_identifier_mode = current;
```
Comment (midenok) [inline mysql-test/suite/versioning/t/vtmd.test]: Check
```
set versioning_identifier_mode= default;
select @@versioning_identifier_mode;
```
Please, follow style convention of `var= value` (i.e. no space before **"="** on assignment).

File: sql/mysqld.h [hunk 10]
```diff
@@ -215,6 +215,13 @@ enum vers_alter_history_enum
   VERS_ALTER_HISTORY_SURVIVE,
   VERS_ALTER_HISTORY_DROP
 };
+
+enum vers_ident_mode_enum
+{
+  VERS_IDENT_MODE_CURRENT= 0,
+  VERS_IDENT_MODE_HISTORICAL_EARLY,
+  VERS_IDENT_MODE_HISTORICAL
```
Comment (midenok) [inline sql/mysqld.h]: Put `VERS_IDENT_MODE_HISTORICAL` before `VERS_IDENT_MODE_HISTORICAL_EARLY`.

File: sql/vtmd.cc [hunk 11]
```diff
@@ -526,6 +526,12 @@ VTMD_table::find_archive_name(THD *thd, String &out)
   TABLE_LIST *table_list= ctx.table_list;
   TABLE_LIST *first_name_resolution_table= ctx.first_name_resolution_table;
   table_map map = vtmd.table->map;
+  vers_ident_mode_enum mode= (vers_ident_mode_enum)thd->variables.vers_ident_mode;
+  const bool range_and_historical=
+      (about.vers_conditions == FOR_SYSTEM_TIME_FROM_TO ||
+       about.vers_conditions == FOR_SYSTEM_TIME_BETWEEN) &&
+      (mode == VERS_IDENT_MODE_HISTORICAL ||
+       mode == VERS_IDENT_MODE_HISTORICAL_EARLY);
```
Comment (midenok) [inline sql/vtmd.cc]: `(mode >= VERS_IDENT_MODE_HISTORICAL)`

File: sql/vtmd.cc [hunk 12]
```diff
@@ -549,10 +555,34 @@ VTMD_table::find_archive_name(THD *thd, String &out)
   {
     if (!select || select->skip_record(thd) > 0)
     {
-      vtmd.table->field[FLD_ARCHIVE_NAME]->val_str(&out);
-      break;
+      if (range_and_historical)
```
Comment (midenok) [inline sql/vtmd.cc]: You use this flag only here. Decrypt the condition.

File: sql/sql_parse.cc [hunk 13]
```diff
@@ -6390,6 +6456,25 @@ static bool execute_sqlcom_select(THD *thd, TABLE_LIST *all_tables)
...
+            table->vers_conditions == FOR_SYSTEM_TIME_AS_OF)
+        {
+          Dynamic_array<String_pair> map;
+          LEX_STRING db = {table->db, table->db_length};
+          LEX_STRING current_name = {table->alias, strlen(table->alias)};
+          LEX_STRING archive_name = {table->table_name, table->table_name_length};
+          if (get_fields_mapping(thd, db, current_name, archive_name, map))
+            return 1;
+
+          for (Item *item= thd->stmt_arena->free_list; item; item= item->next)
```
Comment (midenok) [inline sql/sql_parse.cc]: Why iterating `free_list`? It's list of free items, how they can be used?

File: sql/sql_parse.cc [hunk 14]
```diff
@@ -6362,6 +6362,72 @@ mysql_execute_command(THD *thd)
...
+                               LEX_STRING archive_name,
+                               Dynamic_array<String_pair> &map)
+{
+  Dynamic_array<const char *> current_fields;
+  Dynamic_array<const char *> archive_fields;
+
+  if (get_table_fields(thd, db, current_name, current_fields))
+    return true;
+  if (get_table_fields(thd, db, archive_name, archive_fields))
+    return true;
```
Comment (midenok) [inline sql/sql_parse.cc]: So much needless data shuffle: put fields in 2 arrays, converge 2 arrays into 1, cycle that array (`map_field()`). And that's all inside single call (`execute_sqlcom_select()`). Why do you need all that?

File: sql/sys_vars.cc (hunk 4 above)
Comment (kevgs) [reply sql/sys_vars.cc]: fixed

File: sql/sys_vars.cc (hunk 5 above)
Comment (kevgs) [reply sql/sys_vars.cc]: fixed

File: sql/sys_vars.cc (hunk 6 above)
Comment (kevgs) [reply sql/sys_vars.cc]: fixed

File: mysql-test/suite/versioning/t/vtmd.test (hunk 7 above)
Comment (kevgs) [reply mysql-test/suite/versioning/t/vtmd.test]: fixed

File: mysql-test/suite/versioning/t/vtmd.test (hunk 8 above)
Comment (kevgs) [reply mysql-test/suite/versioning/t/vtmd.test]: fixed

File: mysql-test/suite/versioning/t/vtmd.test (hunk 9 above)
Comment (kevgs) [reply mysql-test/suite/versioning/t/vtmd.test]: fixed

File: sql/mysqld.h (hunk 10 above)
Comment (kevgs) [reply sql/mysqld.h]: fixed

File: sql/vtmd.cc (hunk 11 above)
Comment (kevgs) [reply sql/vtmd.cc]: fixed

File: sql/sql_parse.cc (hunk 13 above)
Comment (kevgs) [reply sql/sql_parse.cc]: This is not a list of freed items but a list of items to free.
Every item got put on this list in `Item::Item()`.

File: sql/sql_parse.cc (hunk 14 above)
Comment (kevgs) [reply sql/sql_parse.cc]: This temporary code will be removed (except for `map_field()`) once `CMMD` is done.

File: mysql-test/suite/versioning/r/vtmd.result [hunk 15]
```diff
@@ -358,6 +358,82 @@ set versioning_alter_history= survive;
...
+create or replace table t (fire int) with system versioning;
+insert into t values (1);
+set @ts= now(6);
+alter table t add column ice int first, drop column fire;
+select fire from t for system_time as of @ts;
+fire
+1
+select ice from t for system_time as of @ts;
+fire
+1
```
Comment (midenok) [inline mysql-test/suite/versioning/r/vtmd.result]: Expected `ice` in resultset.

File: mysql-test/suite/versioning/r/vtmd.result [hunk 16]
```diff
@@ -358,6 +358,82 @@ set versioning_alter_history= survive;
...
+select ice from t for system_time as of @ts order by ice asc;
```
Comment (midenok) [inline mysql-test/suite/versioning/r/vtmd.result]: What's so special about `order by`?

File: mysql-test/suite/versioning/r/vtmd.result (hunk 15 above)
Comment (midenok) [inline mysql-test/suite/versioning/r/vtmd.result]: That's strange: same column is accessed by 2 different names. Expected error for wrong name.

File: sql/sql_parse.cc (hunk 13 above)
Comment (midenok) [reply sql/sql_parse.cc]: Mm... I guess such semantics anyway does not allow to use it.

File: sql/sql_parse.cc (hunk 14 above)
Comment (midenok) [reply sql/sql_parse.cc]: This needless code should be removed now. You already have simple index mapping (see task description).

File: mysql-test/suite/versioning/r/vtmd.result (hunk 16 above)
Comment (kevgs) [reply mysql-test/suite/versioning/r/vtmd.result]: This test shows that `Item` renaming works all items, not just for `select list` ones.

File: sql/sql_parse.cc (hunk 13 above)
Comment (midenok) [reply sql/sql_parse.cc]: Is `stmt_arena` correct under SP, prepared, derived?

File: sql/sql_parse.cc [hunk 17]
```diff
@@ -6362,6 +6362,72 @@ mysql_execute_command(THD *thd)
...
+  }
+
+  return false;
+}
+
+static void map_field(Item_field *item, const Dynamic_array<String_pair> &map)
+{
+  for (size_t i= 0; i < map.elements(); i++)
+  {
+    if (!strcmp(item->field_name, map.at(i).first))
```
Comment (midenok) [inline sql/sql_parse.cc]: What about `item->table_name`, `item->db_name`?

File: sql/sql_parse.cc [hunk 18]
```diff
@@ -6390,6 +6456,25 @@ static bool execute_sqlcom_select(THD *thd, TABLE_LIST *all_tables)
           return 1;
         if (vtmd.exists && vtmd.setup_select(thd))
           return 1;
+
+        if (thd->variables.vers_ident_mode == VERS_IDENT_MODE_CURRENT &&
+            table->vers_conditions == FOR_SYSTEM_TIME_AS_OF)
+        {
+          Dynamic_array<String_pair> map;
+          LEX_STRING db = {table->db, table->db_length};
+          LEX_STRING current_name = {table->alias, strlen(table->alias)};
```
Comment (midenok) [inline sql/sql_parse.cc]: Using `table->alias` for `current_name` is wrong. Consider
```sql
select rrr.x from t for system_time ... as rrr;
```

**Comment by midenok** [comment]
Postponed until 1.12.

**Pull request closed without merging**